pkgs/development/compilers/factor-lang/mk-factor-application.nix

```diff
@@ -0,0 +1,128 @@
+    factor-lang ? if enableUI then flang else factor-no-gui,
+    enableUI ? false,
+    # Allow overriding the deployed vocabulary name. A $vocabName.factor file must exist!
+    vocabName ? finalAttrs.pname or name,
+    # Allow overriding the binary name
+    binName ? lib.last (lib.splitString "/" vocabName),
+    # Extra libraries needed, defaults to standard libraries, set [] to disable
+    extraLibs ? with factor-lang; defaultLibs ++ runtimeLibs,
+    # Extra binaries in PATH, defaults to standard paths, set [] to disable
+    extraPaths ? with factor-lang; defaultBins ++ binPackages,
+    deployScriptText ? ''
+      USING: command-line io io.backend io.pathnames kernel namespaces sequences
+      tools.deploy tools.deploy.config tools.deploy.backend vocabs.loader ;
+
+      IN: deploy-me
+
+      : load-and-deploy ( path/vocab -- )
+          normalize-path [
+              parent-directory add-vocab-root
+          ] [
+              file-name dup reload deploy
+          ] bi ;
+
+      : deploy-vocab ( path/vocab path/target -- )
+          normalize-path deploy-directory set
+          f open-directory-after-deploy? set
+          load-and-deploy ;
+
+      : deploy-me ( -- )
+          command-line get dup length 2 = [
+              first2 deploy-vocab
+          ] [
+              drop
+              "usage: deploy-me <PATH-TO-VOCAB> <TARGET-DIR>" print
+              nl
+          ] if ;
+
+      MAIN: deploy-me
+    '',
+    ...
+  }@attrs:
+  let
+    deployScript = writeText "deploy-me.factor" finalAttrs.deployScriptText;
+  in
+  {
+    inherit
+      enableUI
+      vocabName
+      deployScriptText
+      extraLibs
+      extraPaths
+      binName
+      factor-lang
+      ;
+    nativeBuildInputs = [ makeWrapper ] ++ attrs.nativeBuildInputs or [ ];
+
+    buildInputs = [
+      finalAttrs.factor-lang
+    ] ++ (lib.optional enableUI gdk-pixbuf) ++ attrs.buildInputs or [ ];
+
+    buildPhase =
+      attrs.buildPhase or ''
+        runHook preBuild
+        vocabBaseName=$(basename "${finalAttrs.vocabName}")
+        mkdir -p "$out/lib/factor" "$TMPDIR/.cache"
+        export XDG_CACHE_HOME="$TMPDIR/.cache"
+
+        ${finalAttrs.factor-lang}/bin/factor ${deployScript} "./${finalAttrs.vocabName}" "$out/lib/factor"
+        cp "$TMPDIR/factor-temp"/*.image "$out/lib/factor/$vocabBaseName"
+        runHook postBuild
+      '';
+
+    installPhase =
+      attrs.installPhase or (
+        ''
+          runHook preInstall
+        ''
+        + (lib.optionalString finalAttrs.enableUI ''
+          # Set Gdk pixbuf loaders file to the one from the build dependencies here
+          unset GDK_PIXBUF_MODULE_FILE
+          # Defined in gdk-pixbuf setup hook
+          findGdkPixbufLoaders "${librsvg}"
+          appendToVar wrapperArgs --set GDK_PIXBUF_MODULE_FILE "$GDK_PIXBUF_MODULE_FILE"
+          appendToVar wrapperArgs --prefix LD_LIBRARY_PATH : /run/opengl-driver/lib
```

Comment:
Your suggestion was my initial implementation. A previous reviewer suggested the (newer?) approach to use `appendToVar`. Looking into its definition in [`pkgs/stdenv/generic/setup.sh`](https://github.com/NixOS/nixpkgs/blob/b1d8131eaf3251933e8aa03b5de718451b9e94bc/pkgs/stdenv/generic/setup.sh#L355), it also uses arrays if applicable. So `appendToVar` should be properly preserving command-line arguments as well.

Reply:
I had thought that `appendToVar` only works on string-like Bash variables. Considering that, it might be a matter of design choices.

-   If you want to enable users to specify `makeWrapperArgs` as an attribute even when `__structuredAttrs` is not enabled, and you don't plan to give that non-structured derivation a way to add command-line arguments containing spaces, `appendToVar` is the right way to go.
-   If you want to enable only the structured derivation to specify `makeWrapperArgs` and allow non-structured derivation to specify command-line arguments containing spaces, we could use `makeWrapperArgs+=()`.

As #318614 is merged, the first is the way forward. If it works, let's use `appendToVar` instead.

Reply:
> Although `appendToVar` tries to use arrays to achieve the same goal, it does not seem to work as intended.

What issues with `appendToVar` have you encountered? It seems to work in `nix-shell`:

```
nixpkgs_factor-lang on  HEAD (41a2dde) [$] 
❯ nix-shell --expr "(import ./. { }).hello.overrideAttrs { __structuredAttrs = true; }"
structuredAttrs is enabled
Using versionCheckHook

nixpkgs_factor-lang on  HEAD (41a2dde) [$] via ❄️  impure 
❯ echo "$__structuredAttrs"
1

nixpkgs_factor-lang on  HEAD (41a2dde) [$] via ❄️  impure 
❯ appendToVar foo hi

nixpkgs_factor-lang on  HEAD (41a2dde) [$] via ❄️  impure 
❯ declare -p foo
declare -a foo=([0]="hi")

nixpkgs_factor-lang on  HEAD (41a2dde) [$] via ❄️  impure 
❯ appendToVar foo Alice and Bob

nixpkgs_factor-lang on  HEAD (41a2dde) [$] via ❄️  impure 
❯ declare -p foo
declare -a foo=([0]="hi" [1]="Alice" [2]="and" [3]="Bob")
```

BTW, you may want to use _accumFlagsArray to harvest the flags collected by `appendToVar`.

Reply:
I fixed that using `__strcuturedAttrs`. Thanks.